### PR TITLE
Weakly reference Activity for transaction finished callback

### LINKED ISSUE #2201: ActivityLifecycleIntegration.class would cast mem leak
### Integration

sentry-android

### Build System

Gradle

### AGP Version

5.4.1

### Proguard

Disabled

### Version

6.2.1

### Steps to Reproduce

image

┬───
│ GC Root: Global variable in native code
│
├─ dalvik.system.PathClassLoader instance
│    Leaking: NO (StatServiceImpl↓ is not leaking and A ClassLoader is never leaking)
│    ↓ ClassLoader.runtimeInternalObjects
├─ java.lang.Object[] array
│    Leaking: NO (StatServiceImpl↓ is not leaking)
│    ↓ Object[9097]
├─ com.tencent.stat.StatServiceImpl class
│    Leaking: NO (a class is never leaking)
│    ↓ static StatServiceImpl.r
│                             ~
├─ io.sentry.UncaughtExceptionHandlerIntegration instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 11488 objects
│    ↓ UncaughtExceptionHandlerIntegration.options
│                                          ~~~~~~~
├─ io.sentry.android.core.SentryAndroidOptions instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 11360 objects
│    ↓ SentryOptions.integrations
│                    ~~~~~~~~~~~~
├─ java.util.concurrent.CopyOnWriteArrayList instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 10900 objects
│    ↓ CopyOnWriteArrayList[8]
│                          ~~~
├─ io.sentry.android.core.ActivityLifecycleIntegration instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 10864 objects
│    application instance of com.app.SkApplication
│    ↓ ActivityLifecycleIntegration.appStartSpan
│                                   ~~~~~~~~~~~~
├─ io.sentry.Span instance
│    Leaking: UNKNOWN
│    Retaining 284 B in 8 objects
│    ↓ Span.transaction
│           ~~~~~~~~~~~
├─ io.sentry.SentryTracer instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 10747 objects
│    ↓ SentryTracer.transactionFinishedCallback
│                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
├─ io.sentry.android.core.-$$Lambda$ActivityLifecycleIntegration$1Xa20D8rbI_RfOTCST8-8mCjmDU instance
│    Leaking: UNKNOWN
│    Retaining 13.2 MB in 10732 objects
│    f$1 instance of com.app.activity.StartActivity with mDestroyed = true
│    ↓ -$$Lambda$ActivityLifecycleIntegration$1Xa20D8rbI_RfOTCST8-8mCjmDU.f$1

when i started main activity from launch activity,the launch activity would be leak.

the activityLifecycleIntegration is single.

### Expected Result

launch actvity no leak

### Actual Result

launch actvity leaked

### PR DESCRIPTION
**Pull request by adinauer**
## :scroll: Description

Wrap `Activity` with `WeakReference` in `ActivityLifecycleIntegration` to not strongly hold on to it for the `TransactionFinishedCallback`.

## :bulb: Motivation and Context

Fixes #2201 

## :green_heart: How did you test it?
Emulator

## :pencil: Checklist

- [ ] I reviewed the submitted code
- [ ] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] No breaking changes

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java [hunk 1]
```diff
@@ -167,7 +169,10 @@ private void startTracing(final @NotNull Activity activity) {
                 (Date) null,
                 true,
                 (finishingTransaction) -> {
-                  activityFramesTracker.setMetrics(activity, finishingTransaction.getEventId());
+                  @Nullable Activity unwrappedActivity = weakActivity.get();
+                  if (unwrappedActivity != null) {
+                    activityFramesTracker.setMetrics(activity, finishingTransaction.getEventId());
+                  }
```
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java:175]: I'd log in to the `else` branch, so people know why frames were not attached in this case.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java [hunk 2]
```diff
@@ -146,6 +147,7 @@ private void stopPreviousTransactions() {
   }
 
   private void startTracing(final @NotNull Activity activity) {
+    WeakReference<Activity> weakActivity = new WeakReference<>(activity);
```
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java:150]: Please try to reproduce the issue with and without the fix using LeakCanary.
Upgrade leakcanary to the latest version -> https://github.com/square/leakcanary/releases/tag/v2.9.1

Comment (adinauer) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java:150]: Tried with `2.9.1` but it didn't find anything regardless of `WeakReference` or old implementation.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java (hunk 1 above)
Comment (adinauer) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java]: Updated

Comment (vvgool) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java]: image

is still use activity , replace by unwrappedActivity?

Comment (adinauer) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java]: I think you looked at some old version of the code in this PR. It's using the `unwrappedActivity` now. I unintentionally commited the code I played around with to confirm the fix and then fixed with another commit.

**Review by marandaneto — APPROVED** [review]

**Pull request merged**